includes/class-rc-order.php

```diff
@@ -0,0 +1,166 @@
+        $post_fields = [
+            'accessID'              => $this->api_id,
+            'first_name'            => $this->first_name,
+            'last_name'             => $this->last_name,
+            'email'                 => $this->email,
+            'order_timestamp'       => $this->order_timestamp,
+            'browser_ip'            => $this->browser_ip,
+            'user_agent'            => $this->user_agent,
+            'invoice_amount'        => $this->total,
+            'currency_code'         => $this->currency,
+            'external_reference_id' => $this->order_number,
+            'timestamp'             => time(),
+        ];
+
+        // only add referrer_id if present
+        if ($this->referrer_id != null) {
+            $post_fields['referrer_id'] = $this->referrer_id;
+        }
+
+        // check if we need only specific post fields from the default
+        if ($specific_keys != null && count($specific_keys) > 0) {
+            $new_post_fields = [];
+            foreach($post_fields as $field => $value) {
+                if (in_array($field, $specific_keys)) {
+                    $new_post_fields[$field] = $value;
+                }
+            }
+
+            // only overwrite post fields if at least one key is retreived
+            if ($new_post_fields != null && count($new_post_fields) > 0) {
+                $post_fields = $new_post_fields;
+            }
+        }
+
+        // check if there are additional keys we want to add to the payload
+        if ($additional_keys != null && count($additional_keys) > 0) {
+            $post_fields = array_merge($post_fields, $additional_keys);
+        }
+
+        // sort keys
+        ksort($post_fields);
+
+        return $post_fields;
+    }
+
+    // created this function because PHP's http_build_query function converts 'timestamp' to 'xstamp'
+    private function prepParams(Array $params) {
+        $preppedParams = '';
+        foreach($params as $key => $value) {
+            $preppedParams .= "$key=$value";
+        }
+
+        return $preppedParams;
+    }
+
+    private function generate_request_body($post_fields) {
+        if (!empty($this->secret_key) && !empty($this->api_id)) {
+            $params = [
+                'body' => $post_fields
+            ];
+            $params['body']['signature'] = md5($this->secret_key . $this->prepParams($post_fields));
+
+            return $params;
+        }
+    }
+
+    // https://www.referralcandy.com/api#purchase
+    public function submit_purchase() {
+        $endpoint = 'https://my.referralcandy.com/api/v1/purchase.json';
+
+        if (!empty($this->secret_key) && !empty($this->api_id)) {
+            $response = wp_safe_remote_post(
+                $endpoint,
+                $this->generate_request_body($this->generate_post_fields())
+            );
+        }
+    }
+
+    // https://www.referralcandy.com/api#referral
+    public function process_return() {
```

Comment:
So this is the new feature you talked about, where we invalidate a referral if the purchase is returned?

That's a cool feature actually, we should implement it more widely if it works.

Reply:
yup. i agree with what clark said tho. i think i still lack research on what the repercussions could be with an automated void of the referrals. i will remove this for now and keep a copy for future reference. i think a better solution would be to add an option to void the referral upon refund in the future. 🤔

Reply:
Its strange because its an option we have in the API so it is out there in production but I guess not enough people use it for us to understand it 😅